File: Topic-wise/Prefix Sum/in_place_prefix_sum/in_place_prefix_sum.py
"""
    Given an array A. Construct prefix sum of the array in the given array itself.
"""

class DSA:
    def in_place_prefix_sum(self, arr):
        for idx, val in enumerate(arr[1:], start=1):
            arr[idx] = arr[idx-1] + val
        return arr

dsa = DSA()
arr = list(map(int, input().split()))
orignal = tuple(arr)
dsa.in_place_prefix_sum(arr)
print(f'{arr} is the prefix of array, {orignal}')

File: Topic-wise/Prefix Sum/in_place_prefix_sum/in_place_prefix_sum.rs
/*
    Given an array A. Construct prefix sum of the array in the given array itself.
*/

use std::io;

struct DSA;

impl DSA {
    pub fn in_place_prefix_sum(arr: &mut Vec<i64>) -> &Vec<i64> {
        let n = arr.len();
        for i in 1..n {
            arr[i] = arr[i - 1] + arr[i];
        }
        arr
    }
}

fn main() {
    let mut input = String::new();
    io::stdin()
        .read_line(&mut input)
        .expect("Failed to read line");

    let mut arr: Vec<i64> = input
        .trim()
        .split_whitespace()
        .map(|x| x.parse().expect("Not an integer"))
        .collect();

    let orignal = arr.clone();
    println!(
        "{:?} is the prefix of array, {:?}",
        DSA::in_place_prefix_sum(&mut arr),
        orignal
    );
}
